sdk docs (#734)
flowglad/flowglad: platform/docs/sdks/server.mdx

---
title: "@flowglad/server"
description: "Core server-side SDK for Flowglad integration in Node.js applications"
---

import ServerSetup from '/snippets/setup-server.mdx'

<Info>
  NPM Package: [@flowglad/server](https://www.npmjs.com/package/@flowglad/server)
</Info>

## Overview

The Server SDK is the core package for server-side Flowglad integration. It provides the `FlowgladServer` class that handles all billing operations, customer management, and API communication with Flowglad.

## Installation

<CodeGroup>

```bash npm
npm install @flowglad/server
```

```bash yarn
yarn add @flowglad/server
```

```bash bun
bun add @flowglad/server
```

</CodeGroup>

<ServerSetup />

## Key Features

- **Customer Management**: Find or create customers automatically
- **Subscriptions**: Create, update, and cancel subscriptions
- **Checkout Sessions**: Generate checkout URLs for payments
- **Feature Access**: Check customer access to features

> - **Usage Tracking**: Record usage events for metered billing

Reviewer: ```suggestion
- **Usage Tracking**: Record usage events and track usage credit grants in real time
```

- **Invoices**: Retrieve and manage invoices
- **Payment Methods**: Handle customer payment methods
- **Pricing Model**: Access your pricing model with products, prices, and features

## API Reference

### Constructor

```typescript
new FlowgladServer(options: FlowgladServerOptions)
```

#### Options

```typescript
interface FlowgladServerOptions {
  apiKey?: string // Defaults to FLOWGLAD_SECRET_KEY env var
  baseURL?: string // Defaults to https://app.flowglad.com
  getRequestingCustomer: () => Promise<CustomerInfo | null>

> }

Reviewer: Should we include auth adapter options here as well?

interface CustomerInfo {
  externalId: string
  name: string
  email: string
  metadata?: Record<string, any>
}
```

### Customer Methods

#### `getRequestingCustomerId()`

Get the ID of the current customer.

```typescript
const customerId = await flowgladServer.getRequestingCustomerId()

> ```

Reviewer: I think we can omit this one, in practice they should never need it

> ```

Reviewer: rm, not really used or meant to be used publicly
```suggestion
```

#### `findOrCreateCustomer()`

Find an existing customer or create a new one.

```typescript
const customer = await flowgladServer.findOrCreateCustomer()
```

#### `getSession()`

Get the requesting customer's identity (externalId, name, email) derived from your auth integration.

```typescript
const session = await flowgladServer.getSession()

> ```

Reviewer: rm, not to be used or meant to be used publicly
```suggestion
```

  type: 'subscription', // or 'one_time'
  priceId: 'price_123',
  successUrl: 'https://example.com/success',
  cancelUrl: 'https://example.com/cancel',
  quantity: 1,
  metadata: {
    campaign: 'summer_sale',
  },
})

// Redirect user to checkoutSession.url
```

**Parameters:**

```typescript
interface CreateCheckoutSessionParams {
  priceId: string
  successUrl: string
  cancelUrl: string
  quantity?: number
  outputMetadata?: Record<string, unknown>
  outputName?: string
}
```

#### `createAddPaymentMethodCheckoutSession`

Create a checkout session that collects a payment method for the current customer.

```typescript
const session = await flowgladServer.createAddPaymentMethodCheckoutSession({
  successUrl: 'https://example.com/payment-methods/success',
  cancelUrl: 'https://example.com/payment-methods/cancel',
  targetSubscriptionId: 'sub_123', // optional
})
```

#### `createActivateSubscriptionCheckoutSession`

Create a checkout session that activates a provisioning or imported subscription.

```typescript
const activationSession =
  await flowgladServer.createActivateSubscriptionCheckoutSession({
    targetSubscriptionId: 'sub_123',

>     priceId: 'price_123',

Reviewer: ```suggestion
    priceSlug: 'premium_plan',
```

  amount: number
  priceId: string
  subscriptionId: string
  usageMeterId: string
  transactionId: string
  usageDate?: number
  properties?: Record<string, unknown>
}
```

## Common Patterns

### Feature Gating Middleware

```typescript
import { FlowgladServer } from '@flowglad/server'

export function requireFeature(featureId: string) {
  return async (req, res, next) => {
    const hasAccess = await flowgladServer.checkFeatureAccess(featureId)

    if (!hasAccess) {
      return res.status(403).json({
        error: 'Feature not available',
        message: 'Please upgrade your plan to access this feature',
      })
    }

    next()
  }
}

// Usage
app.get('/api/premium-endpoint', requireFeature('premium_feature'), (req, res) => {
  res.json({ data: 'premium data' })
})
```

### Usage Metering

```typescript
export async function trackAPIUsage(endpoint: string) {
  await flowgladServer.createUsageEvent({
    featureId: 'api_calls',
    quantity: 1,
    usageDate: new Date().toISOString(),
    metadata: {
      endpoint,
    },
  })
}

// Usage
app.get('/api/data', async (req, res) => {
  await trackAPIUsage('/api/data')
  res.json({ data: 'response' })
})
```

> ### Subscription Upgrade Flow

Reviewer: is this how we do upgrades? I think we should perhaps consider removing this section until adjustSubscription is ready for GA?

  }
}
```

### Customer Portal Generation

```typescript
export async function getCustomerPortalData() {
  const billing = await flowgladServer.getBilling()

  return {
    customer: billing.customer,
    activeSubscriptions: billing.subscriptions.filter(
      (s) => s.status === 'active'
    ),
    paymentMethods: billing.paymentMethods,
    recentInvoices: billing.invoices.slice(0, 10),
    features: billing.features,
  }
}
```

## Integration Examples

<Info>
Native constructor params are available for popular auth providers including Supabase Auth, Clerk, Next Auth. Read more about auth adaptors [here](/sdks/auth-adaptors)
</Info>

### With Supabase Auth

```typescript
import { FlowgladServer } from '@flowglad/server'
import { createClient } from '@supabase/supabase-js'

const supabase = createClient(
  process.env.SUPABASE_URL!,
  process.env.SUPABASE_SERVICE_KEY!
)

export const flowgladServer = new FlowgladServer({
  getRequestingCustomer: async () => {
    const { data: { user } } = await supabase.auth.getUser()

    if (!user) return null

    return {
      externalId: user.id,
      name: user.user_metadata.full_name || user.email!,
      email: user.email!,
    }
  },

> })

Reviewer: Consider rewriting to showcase with the supabaseAuth adapter

```

### With Firebase Auth

```typescript
import { FlowgladServer } from '@flowglad/server'
import admin from 'firebase-admin'

admin.initializeApp()

export const flowgladServer = new FlowgladServer({
  getRequestingCustomer: async (context) => {
    const token = context?.req?.headers.authorization?.replace('Bearer ', '')

    if (!token) return null

    try {
      const decodedToken = await admin.auth().verifyIdToken(token)

      return {
        externalId: decodedToken.uid,
        name: decodedToken.name || decodedToken.email!,
        email: decodedToken.email!,
      }
    } catch (error) {
      return null
    }
  },

> })

Reviewer: rewrite with factory method, because `getRequestingCustomer` doesn't get a `context` argument
